### PR TITLE
Allow base tags to specify a base class

### PR DESCRIPTION
**Pull request by nilsvu**
## Proposed changes

As discussed in a call today with @kidder and @wthrowe, this PR allows base tags to specify a `type` that is the base class of its descendants. This type is used by `const_item_type` when no full DataBox tag list is available.

### Types of changes:

- [ ] Bugfix
- [x] New feature
- [ ] Refactor

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### CODE REVIEW

**Review by nilsdeppe — CHANGES_REQUESTED** [review]
Please include me in discussions of such low-level code I wrote going forward.

I'm blocking this until I get an update or have time to read it.

File: tests/Unit/DataStructures/DataBox/Test_BaseTags.cpp [hunk 1]
```diff
@@ -60,6 +60,16 @@ struct ArrayComputeBase : Array<I>, db::ComputeTag {
...
 };
 /// [compute_template_base_tags]
+
+class BaseClass {};
+class Subclass : public BaseClass {};
+
+struct BaseClassTag : db::BaseTag {
+  using type = BaseClass;
+};
+struct SubclassTag : BaseClassTag ,db::SimpleTag {
```
Comment (wthrowe) [inline tests/Unit/DataStructures/DataBox/Test_BaseTags.cpp:70]: Formatting: ' ,' -> ', '

**Review by wthrowe — COMMENTED** [review]
Looks good.  Squash the formatting fix directly.

**Review by moxcodes — APPROVED** [review]
Seems like a pretty good solution.

**Comment by nilsdeppe** [comment]
I don't really like this, it just introduces bugs and it's not at all clear to me what this is trying to solve. Could you please post an example of what the issue you are having is? From talking with @kidder it sounds like you're having trouble with interface compute tags, not DataBox. I have an idea of what would be a fix that doesn't add more bugs, but I would like some code to test with to make sure it resolves your issue. Thanks!

**Comment by nilsvu** [comment]
I don‘t understand what you mean with „introduces more bugs“. By that standard you can‘t add any new features to the code...

Here‘s what this feature is intended for:
I have a class like this:
```cpp
struct ElasticityFluxes {
  using argument_tags = tmpl::list<ConstitutiveRelationBaseTag, StrainTag>;
  using return_tags = tmpl::list<StressTag>;
  static void apply(
    gsl::not_null<tnsr::II<...>*> stress,
    const ConstitutiveRelationBase& constitutive_relation,
    const tnsr::i<...>& strain) { ... }
};
```
The argument_tags here are processed by other  classes to build function signatures, e.g. like this:
```cpp
template <typename... FluxesArgumentTags> // These are retrieved from e.g. ElasticityFluxes::argument_tags
struct NumericalFlux {
  using argument_tags = tmpl::list<FluxesArgumentTags..., /* more tags */>;
  static void package_data(
    const db::const_item_type<FluxesArgumentTags>&... fluxes_args, /* more args */) { ... }
};
```

The full DataBox tags list is not necessarily available everywhere such function signatures need to be built, so base tags are an issue. But often base tags are actually for types that have a base class. Then the base tag can carry the base class type and everything works as expected. Interface tags are one example where this issue is present currently, another that I know of because I wrote it is the elliptic numerical flux.

The example aside, I think it makes sense for base tags to carry the base class type. Then, where you don‘t have access to the concrete subclass type, probably because you don‘t _need_ to, you can use the base tag and still know you‘re working with the base class. What do you think?

**Comment by nilsdeppe** [comment]
If a "feature" only introduces a bug, then I don't think it should be added. As far as I can tell, all this PR does is introduce a bug.

I can't really compile your code example and test it. However, why don't you just not use `const_item_type`? Then everything should compile as far as I can tell. That is,
```cpp
template <typename... Ts>
static void packaged_data(const Ts&... ts)
```
It looks like all you're trying to do is forward anyway, and using variadic templates is quite standard for forwarding arguments.

No, as far as I can tell it makes absolutely no sense for base tags to carry a type. Base tags having nothing to do with dynamic polymorphism and thus they never should carry a type. You happen to be using base tags with dynamic polymorphism, but they're much more general than that.

**Comment by nilsvu** [comment]
What do you mean with all this PR does is introduce a bug? ...

File: src/DataStructures/DataBox/DataBoxTag.hpp [hunk 2]
```diff
@@ -223,17 +226,32 @@ struct item_type_impl {
...
+};
+
+CREATE_HAS_TYPE_ALIAS(type)
+CREATE_HAS_TYPE_ALIAS_V(type)
+
+template <typename Tag>
+struct const_item_type_impl<Tag, NoSuchType,
+                            Requires<is_base_tag_v<Tag> and has_type_v<Tag>>> {
+  using type = tmpl::type_from<Tag>;
+};
```
Comment (nilsdeppe) [inline src/DataStructures/DataBox/DataBoxTag.hpp:243]: This is the bug. You are adding undefined behavior into base tags, and making base tags behave differently if passed into `const_item_type` and `item_type` in a way that makes no sense from a user perspective. It also means that now there is a semi-implicit relation between base tags and simple tags, and their type, that is not a simple relationship. You are *literally breaking all the guarantees of base tags*

**Comment by nilsvu** [comment]
superseded by #2111 and #2112

**Pull request closed without merging**